### PR TITLE
feat: disintegrate block

### PR DESCRIPTION
**Pull request by reginleiff**

### CODE REVIEW

**Comment by nicholaschuayunzhi** [comment]
@reginleiff are you able to come up with a more suitable name for this. Sounds very similar to fragile plugin? I'm sorry cant help much in this regard 😛

**Comment by reginleiff** [comment]
@nicholaschuayunzhi Sorry, but are you referring to the interface `IDisplaceable` or the `Disintegrate` plugin? Just to be sure.

**Comment by teclu** [comment]
I propose DESTRUCTOBLOCK (in caps) as the block name; DESTRUCTION (also in caps) as the Plugin.

**Comment by nicholaschuayunzhi** [comment]
Was talking about the plugin. I like `IDisplaceable` though :-)

File: Assets/ScriptableObjects/Plugin Scripts/IDisplaceable.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using UnityEngine;
+
+public interface IDisplaceable
+{
+
+    bool CanMove();
+
+    bool IsDisplaced();
+
+    bool IsMoveablePlugin();
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/IDisplaceable.cs]: I don't think we should have this.

File: Assets/Scripts/PlayerMouse.cs [hunk 2]
```diff
@@ -79,14 +79,15 @@ void SetNewFaces()
...
+                if (displaceable == null || !displaceable.CanMove()) continue;
+
+                if (!displaceable.IsDisplaced() ||
+                    blockFaceOfNeighbouringBlock == displaceable.GetDisplacedFace())
                 {
-                    BlockFace moveableFace = moveable.IsDisplaced()
-                        ? moveable.GetDisplacedFace()
+                    BlockFace moveableFace = displaceable.IsDisplaced()
+                        ? displaceable.GetDisplacedFace()
                         : blockFaceOfNeighbouringBlock;
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/PlayerMouse.cs]: [1/2]
I realised this check is useless. Lets remove it. Line 86 - 87 already makes it such that `moveableFace` will always be `blockFaceOfNeighbouringBlock`

Edit: Just wanna add that this was my bad 😿

Comment (nicholaschuayunzhi) [reply Assets/Scripts/PlayerMouse.cs]: [2/2]
We can then simply line `86 - 87` by creating another Method in `IDisplaceable`, something along the lines of 
```cs
// returns true if block can be displaced in the face direction provided
bool CanBeDisplaceInFaceDir(BlockFace)
```
Which does the appropriate check for moveable and disintegrate.

At the same time, we don't need `IDisplaceable#isDisplaced()`, `IDisplaceable#CanMove()`, `IDisplaceable#GetDisplacedFace()` any more with this change.

File: Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs [hunk 3]
```diff
@@ -33,13 +33,23 @@ public override void OnFaceClick(BlockFace face)
...
+    public BlockFace GetDisplacedFace()
+    {
+        return _displacedFace;
+    }
+
+    public bool CanMove()
+    {
+        return true;
+    }
+
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs]: This doesnt seem very accurate as a block may not be able to be moved when it is translating.  But see my comments in `PlayerMouse` first.

File: Assets/Scripts/BlockFaceBehaviour.cs [hunk 4]
```diff
@@ -43,17 +43,45 @@ public bool FireRaycastFromFace(float skinToLengthRatio, LayerMask collidableLay
...
+
         }
         return false;
     }
 
+    public GameObject GetRaycastObjectRef(float skinToLengthRatio, LayerMask collidableLayers, BlockFace face)
+    {
+        Vector3 normal = face.GetNormal();
+        Vector3[] skinVertices = GetSkinVertices(skinToLengthRatio, collidableLayers, normal, face);
+
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/BlockFaceBehaviour.cs]: remove the white space here 😅

File: Assets/Scripts/BlockFaceBehaviour.cs [hunk 5]
```diff
@@ -27,10 +27,10 @@ public float GetFaceLength()
         return FaceLength;
     }
 
-    public bool FireRaycastFromFace(float skinToLengthRatio, LayerMask collidableLayers, BlockFace face)
+    public Vector3[] GetSkinVertices(float skinToLengthRatio, LayerMask collidableLayers, Vector3 normal, BlockFace face)
     {
         float centerToSkin = FaceLength * (1 - skinToLengthRatio) / 2;
-        Vector3 normal = face.GetNormal();
+
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/BlockFaceBehaviour.cs]: remove whitespace here

**Review by nicholaschuayunzhi — CHANGES_REQUESTED** [review]
Hi @reginleiff, left some comments to make the code more concise.

File: Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs [hunk 6]
```diff
@@ -33,23 +33,21 @@ public override void OnFaceClick(BlockFace face)
...
-    }
+        if (_isDisplaced)
+        {
+            if (face.Equals(_displacedFace))
+            {
+                return true;
+            }
+            else
+            {
+                return false;
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs]: can reduce nesting by return face.Equals(_displacedFace)

File: Assets/ScriptableObjects/Plugin Scripts/DisintegratePlugin.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+                // fire raycast to grab collidable reference
+                GameObject collidedBlock = _block.GetCollidableObject(_directionOfTravel);
+
+                if (collidedBlock != null)
+                {
+                    Destroy(collidedBlock);
+                }
+            }
+        } else if (!_block.IsTranslating())
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/DisintegratePlugin.cs]: style: newline here

File: Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs [hunk 8]
```diff
@@ -33,13 +33,21 @@ public override void OnFaceClick(BlockFace face)
...
+            }
+            
+        }
```
Comment (nicholaschuayunzhi) [inline Assets/ScriptableObjects/Plugin Scripts/MoveablePlugin.cs]: white space here too

**Review by nicholaschuayunzhi — COMMENTED** [review]
This looks pretty cool. Some important notes, set the `disintegrate` material to the `UVMap` component in the `Distintegrate` block prefab. Otherwise the block will become grass (make sure to save and commit). Also i couldnt pull upwards, heard you were able to?

File: Assets/Scripts/PlayerMouse.cs [hunk 9]
```diff
@@ -79,18 +79,13 @@ void SetNewFaces()
...
-                    blockFaceOfNeighbouringBlock == moveable.GetDisplacedFace())
-                {
-                    BlockFace moveableFace = moveable.IsDisplaced()
-                        ? moveable.GetDisplacedFace()
-                        : blockFaceOfNeighbouringBlock;
-                    blockFaceBehaviour.HighlightFace(blockFaceOfNeighbouringBlock);
-                    _faceMap.Add(blockFaceBehaviour, moveableFace); ;
-                }
+                IDisplaceable displaceable = plugin as IDisplaceable;
+                BlockFace moveableFace = blockFaceOfNeighbouringBlock;
```
Comment (nicholaschuayunzhi) [inline Assets/Scripts/PlayerMouse.cs]: we could just use blockFaceOfNieghbouringBlock.

**Comment by reginleiff** [comment]
> we could just use blockFaceOfNieghbouringBlock.

Hmm, I left it there because it explained why we select it for highlighting, but I guess this removes a sort of redundant line.

> Set the disintegrate material to the UVMap component in the Distintegrate block prefab. Otherwise the block will become grass (make sure to save and commit). 

Needed it to test highlighting because the temporary material can't highlight :( 

> Also i couldnt pull upwards, heard you were able to?

No, I could not pull it up but the highlight would appear on the top even though I was not able to pull it.

**Comment by reginleiff** [comment]
> I propose DESTRUCTOBLOCK (in caps) as the block name; DESTRUCTION (also in caps) as the Plugin.

> Sounds very similar to fragile 

I think the plugins describe the behaviour that it is inserting into the block, so what about `DestructiveProjectile`. `DESTRUCTION` kinda works though its a little vague imo.

**Comment by nicholaschuayunzhi** [comment]
> Needed it to test highlighting because the temporary material can't highlight :(

Cant higlight because its not a texture atlas.

> No, I could not pull it up but the highlight would appear on the top even though I was not able to pull it.

It shouldnt highlight in the first place?

File: Assets/ScriptableObjects/Plugin Scripts/DisintegratePlugin.cs (hunk 7 above)
Comment (nicholaschuayunzhi) [reply Assets/ScriptableObjects/Plugin Scripts/DisintegratePlugin.cs]:
```
}
else if
}
```

**Comment by reginleiff** [comment]
> It shouldnt highlight in the first place?

Okay I got it to work properly 👍 because the scene doesn't update the prefab -_-

**Review by nicholaschuayunzhi — APPROVED** [review]
Nice work

**Pull request merged**